File: backend/src/routes/mcpTools.js
// MCP Tools API Routes
// User-defined MCP tool management endpoints

import express from 'express'
import { mcpToolManager } from '../services/mcpToolManager.js'

const router = express.Router()

/**
 * GET /api/mcp-tools/servers
 * List all loaded MCP servers
 */
router.get('/servers', (req, res) => {
  try {
    const status = mcpToolManager.getStatus()

    res.json({
      success: true,
      servers: status.loadedServers,
      totalTools: status.totalTools
    })
  } catch (error) {
    console.error('[MCP Tools] List servers error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * POST /api/mcp-tools/servers
 * Load a new MCP server
 * Body: { name: string, url: string, transport?: string, bearerToken?: string, headers?: object }
 */
router.post('/servers', async (req, res) => {
  try {
    const { name, url, transport, bearerToken, headers } = req.body

    if (!name || !url) {
      return res.status(400).json({
        success: false,
        error: 'Missing required fields: name and url'
      })
    }

    const tools = await mcpToolManager.loadMcpServer(name, {
      url,
      transport,
      bearerToken,
      headers
    })

    res.json({
      success: true,
      message: `Loaded ${tools.length} tools from ${name}`,
      server: name,
      toolsLoaded: tools.length,
      tools: tools.map(tool => ({
        id: tool.id,
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters
      }))
    })
  } catch (error) {
    console.error('[MCP Tools] Load server error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * GET /api/mcp-tools/servers/:name/tools
 * List all tools from a specific server
 */
router.get('/servers/:name/tools', (req, res) => {
  try {
    const { name } = req.params
    const tools = mcpToolManager.listMcpToolsByServer(name)

    res.json({
      success: true,
      server: name,
      tools: tools.map(tool => ({
        id: tool.id,
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters
      })),
      total: tools.length
    })
  } catch (error) {
    console.error('[MCP Tools] List server tools error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * DELETE /api/mcp-tools/servers/:name
 * Unload an MCP server
 */
router.delete('/servers/:name', async (req, res) => {
  try {
    const { name } = req.params

    await mcpToolManager.unloadMcpServer(name)

    res.json({
      success: true,
      message: `Unloaded server: ${name}`
    })
  } catch (error) {
    console.error('[MCP Tools] Unload error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * GET /api/mcp-tools/tools
 * List all MCP tools from all servers
 */
router.get('/tools', (req, res) => {
  try {
    const tools = mcpToolManager.listMcpTools()

    res.json({
      success: true,
      tools: tools.map(tool => ({
        id: tool.id,
        name: tool.name,
        description: tool.description,
        category: tool.category,
        parameters: tool.parameters,
        server: tool.config.mcpServer
      })),
      total: tools.length
    })
  } catch (error) {
    console.error('[MCP Tools] List tools error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * GET /api/mcp-tools/tool/:toolId
 * Get details of a specific MCP tool
 */
router.get('/tool/:toolId', (req, res) => {
  try {
    const { toolId } = req.params
    const tool = mcpToolManager.getMcpTool(toolId)

    if (!tool) {
      return res.status(404).json({
        success: false,
        error: `Tool not found: ${toolId}`
      })
    }

    res.json({
      success: true,
      tool: {
        id: tool.id,
        name: tool.name,
        description: tool.description,
        category: tool.category,
        parameters: tool.parameters,
        server: tool.config.mcpServer,
        metadata: tool.metadata
      }
    })
  } catch (error) {
    console.error('[MCP Tools] Get tool error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

/**
 * POST /api/mcp-tools/fetch
 * Fetch tools from an MCP server URL (temporary connection)
 * Body: { name: string, url: string, transport?: string, bearerToken?: string, headers?: object }
 */
router.post('/fetch', async (req, res) => {
  try {
    const { name, url, transport, bearerToken, headers } = req.body

    if (!name || !url) {
      return res.status(400).json({
        success: false,
        error: 'Missing required fields: name and url'
      })
    }

    const tools = await mcpToolManager.fetchToolsFromServerUrl(name, {
      url,
      transport,
      bearerToken,
      headers
    })

    res.json({
      success: true,
      server: name,
      tools: tools.map(tool => ({
        id: tool.id,
        name: tool.name,
        description: tool.description,
        parameters: tool.parameters
      })),
      total: tools.length
    })
  } catch (error) {
    console.error('[MCP Tools] Fetch tools error:', error)
    res.status(500).json({
      success: false,
      error: error.message
    })
  }
})

export default router

File: backend-python/src/routes/mcp_tools.py
"""
MCP tools API routes.
"""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..services.mcp_tools import mcp_tool_manager


router = APIRouter(tags=["mcp-tools"])


@router.get("/servers")
async def list_servers() -> JSONResponse:
    try:
        status = mcp_tool_manager.get_status()
        return JSONResponse(
            content={
                "success": True,
                "servers": status.get("loadedServers", []),
                "totalTools": status.get("totalTools", 0),
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.post("/servers")
async def load_server(request: Request) -> JSONResponse:
    body = await request.json()
    name = body.get("name")
    url = body.get("url")
    if not name or not url:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Missing required fields: name and url"},
        )
    try:
        tools = await mcp_tool_manager.load_mcp_server(name, body)
        return JSONResponse(
            content={
                "success": True,
                "message": f"Loaded {len(tools)} tools from {name}",
                "server": name,
                "toolsLoaded": len(tools),
                "tools": [
                    {
                        "id": tool.get("id"),
                        "name": tool.get("name"),
                        "description": tool.get("description"),
                        "parameters": tool.get("parameters"),
                    }
                    for tool in tools
                ],
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.get("/servers/{name}/tools")
async def list_server_tools(name: str) -> JSONResponse:
    try:
        tools = mcp_tool_manager.list_mcp_tools_by_server(name)
        return JSONResponse(
            content={
                "success": True,
                "server": name,
                "tools": [
                    {
                        "id": tool.get("id"),
                        "name": tool.get("name"),
                        "description": tool.get("description"),
                        "parameters": tool.get("parameters"),
                    }
                    for tool in tools
                ],
                "total": len(tools),
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.delete("/servers/{name}")
async def unload_server(name: str) -> JSONResponse:
    try:
        await mcp_tool_manager.unload_mcp_server(name)
        return JSONResponse(content={"success": True, "message": f"Unloaded server: {name}"})
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.get("/tools")
async def list_tools() -> JSONResponse:
    try:
        tools = mcp_tool_manager.list_mcp_tools()
        return JSONResponse(
            content={
                "success": True,
                "tools": [
                    {
                        "id": tool.get("id"),
                        "name": tool.get("name"),
                        "description": tool.get("description"),
                        "category": tool.get("category"),
                        "parameters": tool.get("parameters"),
                        "server": tool.get("config", {}).get("mcpServer"),
                    }
                    for tool in tools
                ],
                "total": len(tools),
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.get("/tool/{tool_id}")
async def get_tool(tool_id: str) -> JSONResponse:
    try:
        tool = mcp_tool_manager.get_mcp_tool(tool_id)
        if not tool:
            return JSONResponse(
                status_code=404,
                content={"success": False, "error": f"Tool not found: {tool_id}"},
            )
        return JSONResponse(
            content={
                "success": True,
                "tool": {
                    "id": tool.get("id"),
                    "name": tool.get("name"),
                    "description": tool.get("description"),
                    "category": tool.get("category"),
                    "parameters": tool.get("parameters"),
                    "server": tool.get("config", {}).get("mcpServer"),
                    "metadata": tool.get("metadata"),
                },
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})


@router.post("/fetch")
async def fetch_tools(request: Request) -> JSONResponse:
    body = await request.json()
    name = body.get("name")
    url = body.get("url")
    if not name or not url:
        return JSONResponse(
            status_code=400,
            content={"success": False, "error": "Missing required fields: name and url"},
        )
    try:
        tools = await mcp_tool_manager.fetch_tools_from_server_url(name, body)
        return JSONResponse(
            content={
                "success": True,
                "server": name,
                "tools": [
                    {
                        "id": tool.get("id"),
                        "name": tool.get("name"),
                        "description": tool.get("description"),
                        "parameters": tool.get("parameters"),
                    }
                    for tool in tools
                ],
                "total": len(tools),
            }
        )
    except Exception as exc:
        return JSONResponse(status_code=500, content={"success": False, "error": str(exc)})

